pkg/ensurance/analyzer/sort/throttle-sort.go

```diff
@@ -0,0 +1,137 @@
+func throttleClassAndPriority(p1, p2 executor.ThrottlePod) int {
+	return int(executor.CompareClassAndPriority(p1.ClassAndPriority, p2.ClassAndPriority))
+}
+
+// throttleUseExtResource compares pod by whether use ext resource
+func throttleUseExtResource(p1, p2 executor.ThrottlePod) int {
+	use1 := utils.TransBool2Uint(p1.UseExtResource)
+	use2 := utils.TransBool2Uint(p2.UseExtResource)
+
+	if use1 == use2 {
+		return 0
+	}
+	if use1 < use2 {
+		return -1
+	}
+	return 1
+}
+
+// throttleExtCpuUsed2Limit compares the partition extcpu usage of extcpu limit
+func throttleExtCpuUsed2Limit(p1, p2 executor.ThrottlePod) int {
+	// if both pod don't use ext resource, then return
+	if p1.UseExtResource == false && p2.UseExtResource == false {
+		return 0
+	}
+
+	p1Ratio := p1.PodCPUUsage / float64(p1.ExtCpuLimit)
+	p2Ratio := p2.PodCPUUsage / float64(p2.ExtCpuLimit)
+
+	if utils.AlmostEqual(p1Ratio, p2Ratio) {
+		return 0
+	}
+	if p1Ratio < p2Ratio {
+		return -1
+	}
+	return 1
+}
+
+// throttleCpuUsage compares the partition extcpu usage of extcpu limit
+func throttleCpuUsage(p1, p2 executor.ThrottlePod) int {
+	var p1usage, p2usage float64
+	// if both pod is PodQOSBestEffort, then compare the absolute usage;otherwise, cmpare the ratio compared with PodCPUQuota
+	if p1.ClassAndPriority.PodQOSClass == v1.PodQOSBestEffort && p2.ClassAndPriority.PodQOSClass == v1.PodQOSBestEffort {
+		p1usage = p1.PodCPUUsage
+		p2usage = p2.PodCPUUsage
+	} else {
+		p1usage = p1.PodCPUUsage * p1.PodCPUPeriod / p1.PodCPUQuota
+		p2usage = p2.PodCPUUsage * p2.PodCPUPeriod / p2.PodCPUQuota
+	}
+	if utils.AlmostEqual(p1usage, p2usage) {
+		return 0
+	}
+	if p1usage < p2usage {
+		return -1
+	}
+	return 1
+}
+
+// Cmp compares p1 and p2 and returns:
+//
+//   -1 if p1 <  p2
+//    0 if p1 == p2
+//   +1 if p1 >  p2
+//
+type cmpThrottleFunc func(p1, p2 executor.ThrottlePod) int
+
+// throttleSorter implements the Sort interface, sorting changes within.
+type throttleSorter struct {
+	pods []executor.ThrottlePod
+	cmp  []cmpThrottleFunc
+}
+
+// Sort sorts the argument slice according to the less functions passed to ThrottleOrderedBy.
+func (ms *throttleSorter) Sort(pods []executor.ThrottlePod) {
+	ms.pods = pods
+	sort.Sort(ms)
+}
+
+// ThrottleOrderedBy returns a throttleSorter that sorts using the cmpThrottleFunc functions, in order.
+// Call its Sort method to sort the data.
+func ThrottleOrderedBy(cmp ...cmpThrottleFunc) *throttleSorter {
+	return &throttleSorter{
+		cmp: cmp,
+	}
+}
+
+// Len is part of sort.Interface.
+func (ms *throttleSorter) Len() int {
+	return len(ms.pods)
+}
+
+// Swap is part of sort.Interface.
+func (ms *throttleSorter) Swap(i, j int) {
+	ms.pods[i], ms.pods[j] = ms.pods[j], ms.pods[i]
+}
+
+// Less is part of sort.Interface.
+func (ms *throttleSorter) Less(i, j int) bool {
+	p1, p2 := ms.pods[i], ms.pods[j]
+	var k int
+	for k = 0; k < len(ms.cmp)-1; k++ {
+		cmpResult := ms.cmp[k](p1, p2)
```

Comment:
it will crash, if the length of cmp is zero

Reply:
Ok, i add a validation here, because it removes from the options, so i can't add it in the options.validate